### PR TITLE
feat(server): add usage-based pricing behind a per-account flag

### PR DESCRIPTION
**Pull request by fortmarek**
## What changed

This adds the usage-based pricing from spec [#92](https://hive.tuist.dev/specs/92) behind a per-account FunWithFlags toggle, `:usage_based_pricing`, following the rollout in spec [#93](https://hive.tuist.dev/specs/93). The flag is off for every account by default. With it off, nothing changes: the nightly sync keeps reporting `remote_cache_hit`, and the usage and billing pages render as today.

With the flag on for an account:

1. **Nightly Stripe sync.** `SyncCustomerStripeMetersWorker` passes `usage_based_pricing: true` to `Billing.customer_meter_values/4`. That replaces `remote_cache_hit` with three meters:
   - `cache_egress_megabytes`
   - `cache_requests`
   - `passing_test_cases`

   Runner, LLM token and idempotency handling are unchanged. Zero values are still dropped.
2. **Usage page.** Two new cards load asynchronously and follow the runner card's layout.
   - **Cache usage.** Widgets for egress (bytes downloaded from the caches), requests, and usage (the combined charge). Clicking a widget switches both the daily chart and the receipt beside it. Usage (the default) charts the charge stacked by meter, and its receipt lists just an egress line and a requests line with the total. Egress and requests chart their own quantity stacked by project, with a legend (Kura traffic that resolved to no project shows as an unknown project), and their receipts go from the gross value, to the Tuist Runners discount, to the included allowance, to the charge, with a pace footnote.
   - **Empty states.** The runner, cache usage, and test insights cards replace the chart and receipts with an empty state when the period has no usage of that feature. The summary widgets stay.
   - **Cache traffic and traffic by region.** These two cards show raw Kura traffic rather than anything billed, so they are hidden for accounts with the flag, and their queries are skipped. Accounts without the flag still see them.
   - **Test insights usage.** Widgets for passing test cases and the charge. Daily value bars are stacked by project, with a legend and a projected series. One receipt goes from passing test cases, to the allowance, to the charge. A second lists failed, skipped and on-runner test cases as free.
3. **Billing page.** Three allowance bars (cache egress, cache requests, passing test cases) replace the 200-hit bar. The estimated next payment is runner billed plus the new meters' billed amounts. The Pro usage table lists each meter that went past its allowance, where the `$0.50` hit row used to be.

4. **Pricing page** (`/pricing`), behind its own flag, `:usage_based_pricing_page`. The per-account flag can't gate a public page, so this one is checked for everyone when the visitor is anonymous and for the signed-in user otherwise, which lets the team preview it before it goes public. With it on, the compare table replaces the Cache and Tests unit rows ($0.50 per unit above 200) with Cache egress (100 GB, then $0.35 per GB), Cache requests (1M, then $0.01 per 1,000), Test Insights (5M passing test cases, then $2 per million) and Selective Testing (unlimited), and the FAQ gains entries on what egress and requests are and on the Tuist Runners discount. Anonymous marketing responses are cached publicly for 60 seconds (`stale-while-revalidate` a day), so flipping the flag takes that long to reach every visitor.

The pricing lives in `Tuist.Billing.UsagePricing` and the ClickHouse reads in `Tuist.Billing.UsageMeters`:

| Meter | Allowance per period | Rate |
|---|---|---|
| Cache egress | 100 GB (decimal) | $0.35 per GB |
| Cache requests | 1,000,000 | $0.01 per 1,000 |
| Passing test cases | 5,000,000 | $2 per million |

### Report-only Stripe config

`stripe.prices.usage_meters` gets the three meter keys with empty Price ids in the staging, canary and production values, and in the dev `mise.toml`. This is the same pattern as the runner meters. A key that is present reports usage, and an empty Price id means nothing is charged. No Stripe objects were created or changed. **Create the three Meters in Stripe before enabling the flag for any account**, otherwise the meter jobs error. `get_plan/1` and subscription items are untouched.

Egress is reported in whole decimal megabytes, floored per reporting window, so meter values and invoice quantities stay small. Stripe refuses `tiers` together with `transform_quantity`, so the Price should be tiered in raw megabytes: the first 100,000 MB free, then 0.035 cents per MB.

## How usage is measured

**Every meter is attributed to the project owner.**
- Kura rows already carry the tenant account as `account_id`.
- `cas_events` and `test_case_runs` are read for the projects whose `projects.account_id` is the account, never by their own `account_id`, which is the uploader.
- Project ids are bound as chunked `Array(Int64)` params, which avoids the ClickHouse 26.3 `http_max_fields` limit.

**Cache egress and requests:**
- **Kura.** `kura_usage_events` with `direction = 'egress'`, deduplicated by `event_id` like `Tuist.Kura.Usage` does. Every artifact kind Kura reports counts: `module`, `xcode`, `gradle`, `reapi` (shown as Bazel), `nx` and `metro`. Any other kind is not counted, shown, or reported.
- **Legacy compilation cache.** `cas_events` with `action = 'download'`. This reads the raw table rather than `cas_events_daily_stats`, because a reporting window can be split mid-day at a subscription or trial boundary.
- **Legacy module cache.** Not counted, since it records no sizes and is being removed.
- **Requests.** Only download requests count, because uploads are free.

**Tuist Runners traffic counts at half.** Traffic counts as runner traffic when its Kura region is a private region in `Tuist.Kura.Regions` (today `scw-fr-par-runners`). `traffic_plane` is always `public`, so it can't be used. The runner share is halved and floored before reporting, so the shared allowance applies to what's left. The receipts show that as a credit line.

**Passing test cases** are read from `test_case_runs_by_project FINAL` by `ran_at`. Test case runs whose test run is linked to a Tuist Runners job are then subtracted. The link is `test_runs.ci_run_id = runner_jobs.workflow_run_id` for the same account, with the job enqueued up to 7 days before the window. Those test case runs come from `test_case_runs_by_test_run FINAL`. Failed and skipped test cases are free. Builds are never read.

**`test_case_runs_daily_stats` is not enough**, for three reasons:
- It has no `test_run_id` or `ci_run_id`, so it can't exclude tests run on Tuist Runners.
- It is keyed by `toDate(inserted_at)`, so it can't serve the sub-day windows the sync produces at boundaries.
- Its `countState` counts every re-inserted row. Flaky corrections re-insert test case runs with a new `inserted_at`, so those runs would be billed twice.

For the same reason, test case runs are windowed by `ran_at`, which stays stable across re-inserts, and read `FINAL`.

**Every new query uses half-open windows** (`>= start AND < end`). `Tuist.Kura.Usage` is inclusive at both ends and still backs the traffic cards shown to accounts without the flag, so it isn't reused here. Tests cover a rollup and a test case run landing exactly on a boundary.

**Billing period.** The usage page passes its selected period: the subscription's `current_period_start/end` when there is one, the calendar month otherwise. The billing page uses the same rule.

**"Billed" vs "Estimated".** An amount is labelled billed only for an account with an active subscription. Every other account, including Air accounts without a Stripe customer, sees the same figures labelled as an estimate. Those accounts add nothing to the billing page's estimated next payment.

## Known gaps, not fixed here

- **Bazel request counts.** Kura counts a whole Bazel `batch_read_blobs` call as one request, so Bazel requests are undercounted compared to other caches.
- **Legacy compilation cache runner traffic.** It can't be told apart from other traffic, so it counts in full.
- **Self-hosted Kura nodes.** They report into `kura_usage_events` like managed nodes, so their traffic is counted.
- **Legacy Gradle and REAPI tables.** The webhook tables (`gradle_cache_events`, `reapi_cache_events`) aren't read. Gradle and Bazel are counted from Kura only.
- **Late arrivals.** Kura rollups or test case runs that arrive after the nightly sync has reported their day are never reported. The error is biased toward undercharging, like the runner meters.
- **Run id collisions.** A GitLab pipeline id or Buildkite build number that equals a runner `workflow_run_id` in the same account would be treated as a runner test run.
- **Empty Prices.** While the Prices are empty, enabling the flag for a subscribed account shows "Billed" amounts and an estimated payment that Stripe won't charge, and the hit meter stops reporting for that account. Turning the flag on is the switch, so fill in the Prices and swap the subscription items first. The item swap is out of scope for this PR.

Out of scope, as staged in spec #93: the subscription item swap, Air free-tier gate changes, emails, and removing the legacy module cache. No new data is stored, so `server/data-export.md` is unchanged.

## Validation

- `mix test` on these files: 276 tests passed.
  - `test/tuist/billing` (including the new `usage_meters_test.exs` and `usage_pricing_test.exs`)
  - `test/tuist/billing_test.exs`
  - `test/tuist_web/live/usage_live_test.exs`
  - `test/tuist_web/live/billing_live_test.exs`
  - `test/tuist/kura/usage_test.exs`
  - `test/tuist/runners/allowance_test.exs`
- `UsageMetersTest` runs against the real ClickHouse and Postgres test databases. It covers:
  - Kura downloads by project, and runner-region detection.
  - Uploads being ignored, and a redelivered rollup counted once.
  - Half-open windows for Kura rollups and test case runs.
  - Compilation cache downloads attributed to the project owner.
  - Test case runs attributed to the project owner and not the uploader.
  - Test case runs split out by runner job.
  - A re-inserted test case run counted once.
- `UsagePricingTest` covers:
  - Exact allowance boundaries, which are charged nothing.
  - Charges past each allowance.
  - Runner halving before the shared allowance.
  - Floored megabytes in meter values.
  - Excluded test statuses and on-runner test cases.
  - Billed only with a subscription.
  - The projection.
- `BillingTest` covers meter values with the flag on (new meters replace the hit meter, unprovisioned meters are dropped) and off (hit meter only, pricing never read).
- `SyncCustomerStripeMetersWorkerTest` covers the flag being passed through. Existing tests were updated for the new option.
- The LiveView tests cover:
  - Cards hidden without the flag.
  - The cache and test receipts for an account without a subscription (estimate) and with one (billed).
  - The cache card reading real Kura rows.
  - The billing page bars, and the next payment with and without a subscription.
- `mix credo` (no issues), `mix format --check-formatted`, and `prettier --check` on `usage.css`.
- `mix gettext.extract` updated only the `.pot` files. Dollar amounts stay outside `dgettext`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: server/lib/tuist/billing/usage_meters.ex [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+  @runner_job_lookback_days 7
+
+  @doc """
+  Cache downloads per UTC day, split by cache and by whether they were served
+  from a Tuist Runners cache region.
+
+  Returns `%{date, cache, runners, bytes, requests}` rows, where `cache` is one
+  of `:module`, `:xcode`, `:gradle`, `:bazel`, or `:other`.
+  """
+  def cache_downloads(account_id, %DateTime{} = period_start, %DateTime{} = period_end) when is_integer(account_id) do
```
Comment (esnunes) [inline server/lib/tuist/billing/usage_meters.ex:36]: One thing I wanted to flag here: I think Xcode cache downloads served by Kura end up counted twice.

On every download, Kura's `DownloadAttribution::commit` (`kura/src/http.rs`) records a `kura_usage_events` rollup (`artifact_kind = "xcode"`, egress). It also calls `record_project_scoped_cache_event` → `enqueue_xcode_download`, which POSTs to `/webhooks/cache`, and `Webhooks.CacheController` writes that into `cas_events`. `kuraController.analytics.enabled` is `true` in `values-managed-common.yaml`, so managed nodes do both. `cache_downloads/3` then adds `kura_downloads` to `compilation_cache_downloads`, and the latter reads every `cas_events` download for the account's projects whatever its `cache_endpoint`.

I reproduced it locally with one Kura rollup plus the matching signed webhook for a single 1.5 GB download: `meter_values/3` returns `cache_download_megabytes: 3000` and `cache_requests: 2`. The same 1 GB download from `scw-fr-par-runners` meters 1500 MB instead of 500 MB, since the `cas_events` copy is always `runners: false`.

Could `compilation_cache_downloads` count only rows whose `cache_endpoint` belongs to the legacy cache (or skip Kura node endpoints)? A test that inserts both rows for the same download would lock it in.

File: server/lib/tuist/billing/usage_meters.ex [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        where: j.account_id == ^account_id and j.workflow_run_id > 0,
+        where: j.enqueued_at >= ^DateTime.add(period_start, -@runner_job_lookback_days, :day),
+        where: j.enqueued_at < ^period_end,
+        select: j.workflow_run_id
+      )
+
+    test_run_ids =
+      from(t in Test,
+        where: fragment("? IN (?)", t.project_id, type(^project_ids, {:array, :integer})),
+        where: fragment("toInt64OrZero(?)", t.ci_run_id) in subquery(workflow_run_ids),
```
Comment (esnunes) [inline server/lib/tuist/billing/usage_meters.ex:176]: It might be worth scoping this match the way `Tuist.Runners.Jobs.ci_scope/2` already does. Right now a test run counts as on-runners when `toInt64OrZero(ci_run_id)` matches any `runner_jobs.workflow_run_id` in the account, with no provider or pipeline check. For Buildkite, `workflow_run_id` is the build number (`runners/buildkite.ex`), which is only unique within a pipeline, and the comment on `ci_scope/2` calls out exactly that cross-link.

I reproduced it locally: a runner job for `acme/ios-pr` build 42, plus a Buildkite test run from `acme/ios-nightly` with `ci_run_id: "42"` and three passing test cases. All three came back as `runners: true`, and `passing_test_cases` reported 0.

I saw run-id collisions in the known gaps. Since Buildkite build numbers restart per pipeline, though, an org with a few pipelines hits this routinely. `runner_jobs.repository` already carries `org/pipeline` for Buildkite, and test runs have `ci_provider`/`ci_project_handle`, so matching on those as `ci_scope/2` does seems like a small change that closes most of it.

File: server/lib/tuist/billing/usage_pricing.ex [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    if total == 0 or Date.after?(first_remaining, last) do
+      []
+    else
+      first_remaining
+      |> Date.range(last)
+      |> Enum.map(&%{date: &1, dollars: total / days_elapsed})
+    end
+  end
+
+  defp project(quantity, period_start, period_end, usage_end) do
```
Comment (esnunes) [inline server/lib/tuist/billing/usage_pricing.ex:212]: A small UI thing, definitely not a blocker. When a closed period is selected on the usage page, the receipts still say "On track for about … this period." For a closed period `usage_end == period_end`, so `project/4` returns the quantity itself and the three pace labels render. Locally, July showed "On track for about 5.0 GB this period." Returning `nil` from `project/4` once `usage_end` has reached `period_end` would hide it, and the projected bars already disappear in that case.

**Review by esnunes — APPROVED** [review]
Thanks for this, Marek. The description is excellent, and the case for reading the raw tables over half-open windows instead of the daily rollups is well made (the re-insert and sub-day-window points are exactly right). Measuring in ClickHouse and reporting through the existing nightly meter sync with report-only Prices, gated per account, mirrors the runner meters and feels like the right way to stage this.

The main thing I'd want sorted before the flag goes on for anyone is that Xcode downloads served by Kura look like they're counted twice: managed Kura nodes write both `kura_usage_events` and `cas_events` for the same download, and both are summed. I also left a note on scoping the Tuist Runners exemption the way `Runners.Jobs.ci_scope/2` already does, and a small one on the pace footnote for closed periods. I reproduced each of these locally against the test ClickHouse and Postgres.

File: server/lib/tuist/billing/usage_meters.ex (hunk 1 above)
Comment (fortmarek) [reply server/lib/tuist/billing/usage_meters.ex:40]: Confirmed and fixed in 3d79756355c. Your reproduction matched what production shows: `cas_events` over one day holds rows from five `cache-*.tuist.dev` hosts and from `kura-…svc.cluster.local:7443` pods, and the Kura ones are the same downloads the rollups already count.

`compilation_cache_downloads/3` now counts CAS events only from endpoints registered in `cache_endpoints`, which is the path Kura publishes no rollup for. Two things the fix has to get right, both of which bit on the first attempt: `cache_endpoints.url` is stored with the `https://` scheme while the webhook stores the bare host, so matching the raw value reads zero rather than erroring, and an empty registry has to short-circuit or Ecto emits `IN ()`.

There is a test with both records present for one download, and I checked it fails without the filter.

Scale today is two projects and about 310 MB a day, so the bill impact is currently small, but it would have doubled for each account as it moved onto Kura.

File: server/lib/tuist/billing/usage_meters.ex (hunk 2 above)
Comment (fortmarek) [reply server/lib/tuist/billing/usage_meters.ex]: Fixed in e243486a6d6. The match now pairs the run id with the project handle instead of using the run id alone.

I could not call `ci_scope/2` here since this is a set-based match over all of an account's jobs rather than one job at a time, and `runner_jobs` carries no provider column. What it does have is `repository`, which is the pipeline handle for Buildkite and the project handle elsewhere, so both sides are compared as one key: `concat(toString(workflow_run_id), '\n', repository)` against `concat(toString(toInt64OrZero(ci_run_id)), '\n', ci_project_handle)`.

Checked against production before changing it, over two days: 829 test runs match a runner job by run id, and every one also matches by handle, so the tightening costs nothing that matches today. A test covers a second pipeline reusing the same build number, and it fails without the handle in the key.

One thing it still cannot do is `ci_scope/2`'s GitLab host check, since `runner_jobs` has no host to compare against.

File: server/lib/tuist/billing/usage_pricing.ex (hunk 3 above)
Comment (fortmarek) [reply server/lib/tuist/billing/usage_pricing.ex:245]: Fixed in e243486a6d6. `project/4` now returns `nil` once `usage_end` has reached `period_end`, so a closed period drops the three pace footnotes the way it already dropped the projected bars.

**Comment by pepicrft** [comment]
Ran a deep review over the branch and validated each finding against the code. Grouping by strength.

**Confirmed**

1. `server/lib/tuist/billing.ex:808` `switch_to_usage_based_pricing/1` collapses `apply_usage_meter_items`' `:unchanged` into `{:ok, :switched}` via the `with`. The daily worker then logs `"Switched account #{id} to usage-based pricing: switched"` on every convergence pass for every already-migrated account. We lose the trace of when an account was actually first switched, and ops observability drowns.

2. `server/lib/tuist/billing.ex:626` `pick_enterprise_flat_price/3` falls back to `hd(candidates)` when no configured price matches the target currency, with no log. A misconfigured `enterprise.flat_monthly` catalog then throws Stripe's `All items must have pricing in the same currency` and the traceback points at Stripe rather than at the config gap. One `Logger.warning` on the fallback would save the debugging round.

**Real, worth fixing before the wider rollout**

3. `server/lib/tuist/billing.ex:756` `hold_usage_based_pricing_for_existing_subscriptions/0` filters to `status in ["active", "trialing"]`. Pro subs in `past_due`, `unpaid`, or `paused` are not per-actor-disabled. On reactivation to `active`, the sub still carries the legacy usage Price, but the flag is now (globally) on, so between reactivation and the next `SwitchUsageBasedPricingWorker` tick the sync worker stops reporting `remote_cache_hit` and starts emitting meter events nothing on the sub bills. Same-day zero-billing gap.

4. `server/lib/tuist/billing.ex:762` `{:ok, false} = FunWithFlags.disable(...)` raises `MatchError` on the first `{:error, _}` (transient Redis hiccup). The `Enum.map` aborts, so the operator gets a partial hold and, if they flip the global gate without re-running, the unheld tail is exposed to the very gap the hold exists to prevent. `case`-plus-collect degrades gracefully.

5. `server/lib/tuist/billing/workers/switch_usage_based_pricing_worker.ex:33` `Enum.map(&switch/1)` does two synchronous Stripe calls per Pro account with no jitter, no `max_concurrency`, no per-account child job. A `for_all` flip on `:usage_based_pricing_switch` blasts through Stripe's write budget, retries triple the fan-out, and accounts oscillate between switched and un-switched across daily runs until enough attempts succeed. `Task.async_stream` with a modest concurrency, or per-account child jobs, cleans it up.

**Softer, mostly hardening**

6. `server/lib/tuist/billing.ex:117` `customer_meter_values/4` under `usage_based_pricing: true` only checks env-level meter provisioning, never that the sub actually carries the meter Prices. If anyone toggles `:usage_based_pricing` for one account via the FunWithFlags UI without going through `switch_to_usage_based_pricing/1`, meter events fire but nothing on the sub bills them, and `remote_cache_hit` is skipped for the cycle. A MapSet check against the sub's item Prices makes it self-defending.

7. `server/lib/tuist/billing/usage_pricing.ex:267` `test_totals/1` reduce has no fallthrough head. Currently masked by ClickHouse's `Enum8('success','failure','skipped')` constraint, but the moment a new variant lands (or a synthetic row shows up in tests) this raises `FunctionClauseError` and the usage page blows up. A `_row, totals -> totals` catch-all is cheap insurance.

8. `server/lib/tuist/billing/usage_meters.ex:193` `runner_job_keys` depends only on `account_id`, but the subquery is rebuilt inside every project-id chunk. Accounts over 5,000 projects re-scan `runner_jobs` per chunk. Materialise the keys once, splice as a literal. (Note: `runner_jobs` is a ClickHouse table, not Postgres, so it's CH work amplification, not a PG round-trip.)

9. `server/lib/tuist/billing.ex:617` `enterprise_subscription_items` fans out `1 x Customer.retrieve + up to N x Price.retrieve` on every upgrade, plus a fallback loop when the customer has no pinned currency. Reads are small at current enterprise volume, but the currency could just as well come from a local `enterprise.currencies` map keyed by price id. Tech debt.

10. `server/priv/repo/migrations/20260917170000_migrate_invalid_account_handles.exs:30` interpolates `@valid_handle` into the SQL rather than binding it. Safe today; a future edit that introduces an apostrophe breaks the migration silently at compile time. Binding via `repo.query!(sql, [@valid_handle, @max_length])` removes the class. This one is from PR #13358, only in the diff because the branch predates its merge on `main`.

Happy to break the confirmed and real-worth-fixing ones out into separate commits on this branch if that's easier to review.

**Review by pepicrft — APPROVED** [review]

File: server/lib/tuist/billing/usage_meters.ex (hunk 1 above)
Comment (fortmarek) [reply server/lib/tuist/billing/usage_meters.ex:40]: Fixed in 3d79756355c: CAS events are now counted only from endpoints registered in `cache_endpoints`, with a test covering both records for one download.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: server/lib/tuist/billing/usage_meters.ex (hunk 2 above)
Comment (fortmarek) [reply server/lib/tuist/billing/usage_meters.ex]: Fixed in e243486a6d6: the match now pairs the run id with the project handle, verified against production where all 829 run-id matches over two days also match by handle.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: server/lib/tuist/billing/usage_pricing.ex (hunk 3 above)
Comment (fortmarek) [reply server/lib/tuist/billing/usage_pricing.ex:245]: Fixed in e243486a6d6: `project/4` returns `nil` once `usage_end` reaches `period_end`, so a closed period drops the pace footnotes.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by fortmarek** [comment]
Worked through all ten. Three fixed in 878574a1561, the rest I'm pushing back on with reasons.

**Fixed**

1. **`:unchanged` collapsed into `:switched`.** Right, and the consequence is worse than noise: since the worker converges daily over every account the flag covers, the line recording when an account actually moved is buried under a daily repeat forever. The outcome is carried through now and only a real switch logs.

3. **Hold missed recoverable statuses.** Agreed, and the window is wider than a day: an unheld subscription that recovers is only switched if `:usage_based_pricing_switch` covers it, so it can sit reporting meters nothing bills. The hold now covers every status a subscription can return to `active` from. Terminal ones stay out on purpose: an account that resubscribes after cancelling gets a new subscription, and new subscriptions belong on the new pricing. (Production carries 50 active and 15 canceled Pro subscriptions today, and none in the recoverable states, so this is prevention rather than a live gap.)

4. **`{:ok, false} =` raising.** Agreed. It now answers with `%{held:, failed:}` and keeps going, with a test that one failed gate does not stop the rest.

**Not changing**

5. **Stripe fan-out in the worker.** 50 active Pro subscriptions, two calls each, sequential: roughly 100 requests spread over tens of seconds against a ~100/second budget. Sequential *is* the throttle at this size, and per-account child jobs would be machinery for a one-time migration. Worth revisiting if the flag ever covers hundreds of subscriptions. One correction: accounts cannot oscillate, because the switch only ever adds meter items and removes the usage Price, so a retry converges rather than reverting.

6. **`customer_meter_values/4` not checking the subscription's items.** This is the documented contract rather than an oversight: `switch_to_usage_based_pricing/1` enables the flag itself, so the pairing holds unless someone sets the gate by hand in the FunWithFlags UI. Defending against that costs a Stripe read per customer per night, every night, forever. It is called out under "Prices before the switch" in the description.

7. **`test_totals/1` fallthrough.** Deliberately absent. `status` is a ClickHouse `Enum8` of exactly those three values, so a new variant means the schema changed, and for a billing meter I would rather that raise than silently drop a status from the count. A catch-all would turn an undercount into the quiet kind.

8. **`runner_job_keys` rebuilt per chunk.** Real above 5,000 projects, but the largest account in production has 108, so there is always exactly one chunk and the subquery is built once. Happy to hoist it if an account ever approaches the chunk size.

2, 9, 10. **`pick_enterprise_flat_price/3`, `enterprise_subscription_items`, the handle migration.** None are this branch's code; 2 and 9 are on `main` and 10 came from #13358, which the branch picked up by merging `main`. Better as their own PRs than smuggled into this one.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
